### DISCUSSION TITLE
Is it impossible to attach user information to the Ast node?

### DISCUSSION THREAD

**Discussion by negahama** [eclipse-langium/langium#1652] [Q&A] [answered]
Hello...
In my DSL, I want to infer the type of variable but I don't know how to save the type information of variables. 

What I want is like this
```
var a = 0
var s = "s"

// instead of
var a: number = 0
var s: string = "s" 
```

I can infer the type of `a`, `s` by using literal `0`, `"s"`.
The problem is that I can't store the type information of variables.
Do I need to have the table to keep all variables's type?
Isn't there some more smart way? 

Is it impossible to attach some user information(e.g type information) to the Ast node(`a`, `s`)?
If yes, let me know any tip to implement this feature.

Thank you in advance...

**Comment by msujew** [comment, marked as answer]
Hey @negahama,

we usually store this information on a `DocumentCache` or `WorkspaceCache` and simply use `<AstNode, Type>` for the type arguments. That way you don't need to worry about cache invalidation. See also the work in progress PR over at https://github.com/eclipse-langium/langium-website/pull/241 for usage details.

**Reply by negahama** [reply]
Great... thanks a lot
